cg.go

```diff
@@ -0,0 +1,162 @@
+package optimize
+
```

Comment:
All these exported structs need comments. I'm not familiar with CGVariant, so I personally would appreciate some explanation. Up to you what the appropriate level is, as we obviously aren't going to write a book in the comments.